src/common/radix_partitioning.cpp

```diff
@@ -0,0 +1,454 @@
+	D_ASSERT(radix_bits_2 <= sizeof(hash_t) * 8);
+	switch (radix_bits_2) {
+	case 1:
+		return OP::template Operation<radix_bits_1, 1>(std::forward<ARGS>(args)...);
+	case 2:
+		return OP::template Operation<radix_bits_1, 2>(std::forward<ARGS>(args)...);
+	case 3:
+		return OP::template Operation<radix_bits_1, 3>(std::forward<ARGS>(args)...);
+	case 4:
+		return OP::template Operation<radix_bits_1, 4>(std::forward<ARGS>(args)...);
+	case 5:
+		return OP::template Operation<radix_bits_1, 5>(std::forward<ARGS>(args)...);
+	case 6:
+		return OP::template Operation<radix_bits_1, 6>(std::forward<ARGS>(args)...);
+	case 7:
+		return OP::template Operation<radix_bits_1, 7>(std::forward<ARGS>(args)...);
+	case 8:
+		return OP::template Operation<radix_bits_1, 8>(std::forward<ARGS>(args)...);
+	case 9:
+		return OP::template Operation<radix_bits_1, 9>(std::forward<ARGS>(args)...);
+	case 10:
+		return OP::template Operation<radix_bits_1, 10>(std::forward<ARGS>(args)...);
+	default:
+		throw InternalException("TODO");
+	}
+}
+
+template <class OP, class RETURN_TYPE, typename... ARGS>
+RETURN_TYPE DoubleRadixBitsSwitch1(idx_t radix_bits_1, idx_t radix_bits_2, ARGS &&...args) {
+	D_ASSERT(radix_bits_1 <= sizeof(hash_t) * 8);
+	switch (radix_bits_1) {
+	case 1:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 1>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 2:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 2>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 3:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 3>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 4:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 4>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 5:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 5>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 6:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 6>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 7:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 7>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 8:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 8>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 9:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 9>(radix_bits_2, std::forward<ARGS>(args)...);
+	case 10:
+		return DoubleRadixBitsSwitch2<OP, RETURN_TYPE, 10>(radix_bits_2, std::forward<ARGS>(args)...);
+	default:
+		throw InternalException("TODO");
+	}
+}
+
+unique_ptr<idx_t[]> RadixPartitioning::InitializeHistogram(idx_t radix_bits) {
+	auto result = unique_ptr<idx_t[]>(new idx_t[1 << radix_bits]);
+	memset(result.get(), 0, (1 << radix_bits) * sizeof(idx_t));
+	return result;
+}
+
+struct UpdateHistogramFunctor {
+	template <idx_t radix_bits>
+	static void Operation(const UnifiedVectorFormat &hash_data, const idx_t count, const bool has_rsel,
+	                      idx_t histogram[]) {
+		using CONSTANTS = RadixPartitioningConstants<radix_bits>;
+
+		const auto hashes = (hash_t *)hash_data.data;
+		if (has_rsel) {
```

Comment:
I think we can just infer this from `sel->data()`? Plus part of the reason I made the change to use `nullptr` to mean identity was it seemed that the branch prediction from the inlining of `get_index` performed this optimization at runtime. Did you see a performance degradation?

Reply:
I removed the histogram code, therefore this is resolved